Breeding-Insight/bi-api
# [BI-1204] Migrate germplasm

# Description

A java migration to migrate the germplasm in BreedBase. 

Key features of Java migrations: 
1. Java migrations are not applied with the flyway commands, only when you run the micronaut application. 
2. Added an `outOfOrder` config to micronaut flyway so that java migrations can be run after db migrations even if they have a version less that the current version of the db. 

Key features of Migration:
1. Because there is no "rollback" for brapi, the script checks if a germplasm has already been updated before making the updates. This way, if the migration fails we can run it again without continuing to append accession numbers onto the germplasm name. A germplasm is considered already updated if it:
  - Has an accession number
  - Has a germplasm name with the program key and accession number
  - Has an external reference with a germplasm UUID
2. If a program has germplasm to update and If there is a not a program key an error will be thrown and the script will not complete successfully. 
3. After updating a germplasm, the response is checked to see that it was updated. This was added because without the sgn 1204 fix BreedBase will return a 200 without updating the germplasm.
3. The following fields are updated:
  - accessionNumber
    a. Is queried from the germplasm sequence
  - seedSource
    a. If germplasm does not have a seedSource it is set to "Unknown"
  - createdBy->userId and createdBy->userName
    a. Is set to values from the BI system user
  - breedingMethodId and breedingMethodName
    a. Is mapped from breedbase values or set to "Unknown" if one doesn't exist, or a match can't be found. 
  - externalReferences: referenceSource = breeding-insight.net & referenceID = germplasm UUID
    a. Random UUID is generated for the germplasm ID
  - Germplasm name is formatted to `<germplasm name> [<program key>-<accession number>]`
  - Display name is set to the original germplasm name.
4. BreedBase doesn't accept pedigree on PUT, so we don't need to worry about updating the parent references because of the germplasm name changes. 

# Dependencies

BI-1204 of sgn (https://github.com/Breeding-Insight/sgn/pull/47)
develop of bi-web

# How to Test
1. If you don't have a BreedBase program set up already, you can go into the `application.yml` and remove this line `- classpath:org.breedinginsight.db.migration` so the java migration won't run. Then start up the application and create a Breedbase program. Add that line back in after creating your program. 
2. Use insomnia to POST the germplasm in this file to `/brapi/v2/germlasm` (You'll need to change the program id. Get it from bidb)
[1204-germplasm.txt](https://github.com/Breeding-Insight/bi-api/files/7693549/1204-germplasm.txt)
3. Go to Breedbase and upload the following pedigree file 
[BI-1204-pedigree.txt](https://github.com/Breeding-Insight/bi-api/files/7693554/BI-1204-pedigree.txt)
4. Use insomnia to `GET /brapi/v2/germplasm` and check the germplasm. 
5. Start the biapi application. Check that the application starts up ok. 
6. Use insomnia to `GET /brapi/v2/germplasm` and check that the germplasm have been successfully modified.
